### PR TITLE
Add backend weight round robin select

### PR DESCRIPTION
**Pull request by Sherlock-Holo**
## add
- interface `Selector`
- `selector` implement `RandomSelector`
- `selector` implement `WeightRoundRobinSelector`

## reason
`dns-over-https` allow us to set multi backends. However, sometimes some backends will be unavailable, we need a algorithm to reduce probability of thems. WRR algorithm is a good way.

### CODE REVIEW

**Comment by m13253** [comment]
Thank you for the major contribution!

I guess it provides a solution for issue #24.
I will test it when I reach home. If it builds on all platforms and worked well, I will merge it.

**Comment by Sherlock-Holo** [comment]
Actually I don't add any new dependenices, so it should work well for all paltforms as in the past

**Comment by Sherlock-Holo** [comment]
I thinks we should let user know how to set upstream weight legitimately, and we may need to talk about:

1. if upstream timeout, how much we should reduce the weight
2. if upstream returns 5xx error, how much we should reduce the weight

1 and 2 are different, if upstream timeout, usually upstream will be unavailable for a long time, but if it returns 5xx, maybe it just a lot of users query it, it may return to normal quickly

**Comment by m13253** [comment]
We could set the penalty increase exponentially (like in power if around 1.6), but switch to linear after some number of fails, and cap it with a maximum value.

And we always make sure we have at least one server to query even if all servers seems unavailable.

**Comment by Sherlock-Holo** [comment]
yes and we may can learn from some load b balance stuck as traefik

File: doh-client/doh-client.conf [hunk 1]
```diff
@@ -7,40 +7,52 @@ listen = [
...
+
+# CloudFlare's resolver, bad ECS, good DNSSEC
+[[upstream.upstream_google]]
+url = "https://1.1.1.1/dns-query"
+weight = 50
+
+# Google's experimental resolver, good ECS, good DNSSEC
+[[upstream.upstream_ietf]]
+url = "https://dns.google.com/experimental"
+weight= 50
```
Comment (m13253) [inline doh-client/doh-client.conf]: Please add a space here and on the following lines.

File: doh-client/doh-client.conf [hunk 2]
```diff
@@ -76,7 +88,7 @@ passthrough = [
     "time.windows.com",
 ]
 
-# Timeout for upstream request
+# Timeout for upstream request, unit is second
```
Comment (m13253) [inline doh-client/doh-client.conf]: "Timeout for upstream request in seconds"

File: .gitignore [hunk 3]
```diff
@@ -12,3 +12,5 @@
 
 # Project-local glide cache, RE: https://github.com/Masterminds/glide/issues/736
 .glide/
+
+.idea/
```
Comment (m13253) [inline .gitignore]: Would you please add a line ending here?

**Comment by m13253** [comment]
Please make some very small changes before I merge :-)

File: .gitignore (hunk 3 above)
Comment (Sherlock-Holo) [reply .gitignore]: Done

File: doh-client/doh-client.conf (hunk 1 above)
Comment (Sherlock-Holo) [reply doh-client/doh-client.conf]: Add space now

**Comment by Sherlock-Holo** [comment]
The `doh-client.conf` actually is a toml config file, could we change the name suffix to `toml`? some text editor may only check name suffix, they won't highlight the content because they don't know what it is, such as Goland :-(

**Comment by Sherlock-Holo** [comment]
I think I should rewrite the `Selector` interface because it doesn't provide a way to reduce weight in real time. I think reducing weight in real time is necessary.

**Comment by m13253** [comment]
> The `doh-client.conf` actually is a toml config file, could we change the name suffix to `toml`? some text editor may only check name suffix, they won't highlight the content because they don't know what it is, such as Goland :-(

Actually I didn't intend to use a complex format like TOML to represent our configuration files.
What I expect is something like INI or systemd's configuration files. They are just lucky to be compatible with TOML.

Therefore the configuration is actually a simple format (until this pull request) which actually doesn't need any syntax highlighting.

Additionally systemd's configuration files also uses different suffixes (.conf, .service, .target, .socket, .etc). I think Goland is to blame for not recognizing this.

Finally, if we break the compatibility of the configuration files, we need to bump the version number.

**Comment by m13253** [comment]
> I think I should rewrite the `Selector` interface because it doesn't provide a way to reduce weight in real time. I think reducing weight in real time is necessary.

I would really appreciate for your contributions!
I am sorry that I am too busy these months.

**Comment by Sherlock-Holo** [comment]
This PR add `Selector` interface and wrr selector implement, so user should write every upsteam's weight, use `TOML` syntax will reudce compatible break. So we have to bump the major version number ^_^

**Comment by Sherlock-Holo** [comment]
I rewrite `Selector` interface and do a simple test in my PC, the wrr selector works well

**Comment by m13253** [comment]
I tested on my PC.

Thank you for the new deadline detection code.

But one more question: should we make `Client.Selector` private or public? I guess we don't need to share it with other packages?

**Comment by Sherlock-Holo** [comment]
it should be private, I will change it tomorrow

Star Brilliant <notifications@github.com> 于 2019年3月7日周四 23:59写道：

**Comment by Sherlock-Holo** [comment]
Change it pivate now

**Comment by m13253** [comment]
> Change it pivate now

Thank you for the change!

Shall I make the last 2 changes for you, or you will do the changes before I merge?

**Comment by m13253** [comment]
> Also, I hope all `weight= 50` can be replaced by `weight = 50`, `Url` can be replaced by `URL`.

**Comment by Sherlock-Holo** [comment]
- `weight= 50` fixed
- `Url` now is `URL`

**Comment by Sherlock-Holo** [comment]
I upgrade `Selector`, now can report upstream works well. I add it because when I run it on our school DNS server, after reading logs, it seems wrr selector "drop" some upstreams because wrr selector reduce their weight too fast but icrease weight too slow when they are available, so we should increase weight when upstreams are available in real time

**Pull request merged**